iree/compiler/Dialect/Stream/IR/test/tensor_folding.mlir

```diff
@@ -1,5 +1,31 @@
 // RUN: iree-opt -split-input-file -canonicalize %s | iree-opt -split-input-file | IreeFileCheck %s
 
+// CHECK-LABEL: @FoldTensorImportOp
+func @FoldTensorImportOp(%arg0: !stream.resource<external>, %arg1: index) -> !stream.resource<external> {
+  // CHECK-NOT: stream.tensor.import
+  // CHECK-NOT: stream.tensor.export
+  // CHECK: return %arg0 : !stream.resource<external>
+  %c20 = arith.constant 20 : index
+  %0 = stream.tensor.export %arg0 : tensor<?x5xf32>{%arg1} in !stream.resource<external>{%c20} -> !hal.buffer_view
+  %1 = stream.tensor.import %0 : !hal.buffer_view -> tensor<?x5xf32>{%arg1} in !stream.resource<external>{%c20}
+  return %1 : !stream.resource<external>
+}
+
+// -----
```

Comment:
unlike the hal.tensor.cast which carries the encoding in the type here we have the encoding separate - so it's always safe to fold export->import even if the encoding and dims don't match (we assume a correct program on input)

example:
```mlir
  %0 = stream.tensor.export %arg0 : tensor<?x5xf32>{%arg1} in !stream.resource<external>{%c20} -> !hal.buffer_view
  %1 = stream.tensor.import %0 : !hal.buffer_view -> tensor<1x?x5xf32>{%arg1} in !stream.resource<external>{%c20}
```
here even though the exported tensor is ?x5xf32 the imported tensor of 1x?x5xf32 is compatible. I don't think we need to even check the shape/dims at all, just the resource type/size

Reply:
Thanks, that answered a few questions I had... but it also raised a few more :P

1. The import->export case could be a transfer from one device to another, so I think we can't fold across different affinities. However, for the export->import case, would that always be [one of multiple devices] -> [the only host]? If so, I think we _could_ safely fold with different affinities there.
2. Do we need to check the resource size? How is that different from the encoding and shape in this context?

Reply:
export->import could also change devices (both ops have affinities, and those affinities could differ); the semantics on the import/export ops is that the affinities are assertions though instead of transfers (exporting to affinity A just says that future accesses from an A-compatible device should work). It's all a bit fuzzy and I think saying if they don't match we don't fold today is conservative and safe (at least, it keeps the issues isolated to import/export).

The encoding and shapes can differ while sizes are the same - in the above example, a reshape turned ?x5 into 1x?x5 - same size, different shape (could be ?x?{%a, %b} to ?x?{%b, %a}, etc). The sizes need to be tracked consistently and if they differ that's actually invalid IR we could fail verification on - it's just hard (read: expensive) to do verifiers that look through other ops/across calls/etc. By construction they should always be the same as things like reshapes guarantee they don't change storage size and just metadata (the encoding).

Reply:
(that is to say for 2 we could check that they match and ignore if they don't)

Reply:
👍 added back a check for sizes matching.